modules/lfs/http_client.go

```diff
@@ -134,71 +137,84 @@ func (c *HTTPClient) performOperation(ctx context.Context, objects []Pointer, dc
 		return fmt.Errorf("TransferAdapter not found: %s", result.Transfer)
 	}
 
+	errGroup, groupCtx := errgroup.WithContext(context.Background())
 	for _, object := range result.Objects {
-		if object.Error != nil {
-			log.Trace("Error on object %v: %v", object.Pointer, object.Error)
-			if uc != nil {
-				if _, err := uc(object.Pointer, object.Error); err != nil {
-					return err
-				}
-			} else {
-				if err := dc(object.Pointer, nil, object.Error); err != nil {
-					return err
-				}
-			}
-			continue
-		}
+		func(groupCtx context.Context, object *ObjectResponse, dc DownloadCallback, uc UploadCallback, transferAdapter TransferAdapter) {
+			errGroup.Go(func() error {
+				err := performSingleOperation(groupCtx, object, dc, uc, transferAdapter)
+				return err
+			})
+		}(groupCtx, object, dc, uc, transferAdapter)
+	}
 
+	// only the first error is returned, preserving legacy behavior before concurrency
+	return errGroup.Wait()
+}
+
+// performSingleOperation performs an LFS upload or download operation on a single object
+func performSingleOperation(ctx context.Context, object *ObjectResponse, dc DownloadCallback, uc UploadCallback, transferAdapter TransferAdapter) error {
+	if object.Error != nil {
+		log.Trace("Error on object %v: %v", object.Pointer, object.Error)
 		if uc != nil {
-			if len(object.Actions) == 0 {
-				log.Trace("%v already present on server", object.Pointer)
-				continue
+			if _, err := uc(object.Pointer, object.Error); err != nil {
+				return err
 			}
+		}
+		err := dc(object.Pointer, nil, object.Error)
```

Comment:
The changed logic doesn't seem right.

Before: `if uc != nil { uc } else { dc }`

Now: `if uc != nil { uc }  always dc`

Reply:
true, my preference would be to breakout uc/dc into two separate functions and have the switch logic separated out rather than checked multiple times throughout this function. I'll update this branch with that change

Reply:
actually, on closer inspection, the behavior does not look to have changed, this is all within the existing `object.Error` block. It's probably easier to look at the two functions side-by-side, as github diff is showing these as my changes when really it's the same function as before

Reply:
^ still down to refactor this existing code to make it more legible though if you have appetite for that kind of review

Reply:
Maybe adding some comments could help understanding.

Reply:
commented this existing code from main inline, two things:
1) it'd probably be helpful to review this [pr with whitespace disabled](https://github.com/go-gitea/gitea/pull/32369/files?diff=unified&w=1) , as a lot of this code is identical, but outdented due to be in its own function instead of nested in the super function
2) this existing code has a lot of if/else where if the batch api operation is 'upload', it performs one task, and if it's NOT 'upload' it's assumed to be 'download'. Technically this is correct, as the [batch api](https://github.com/git-lfs/git-lfs/blob/main/docs/api/batch.md#requests) only supports those two operation field values, however if I were to refactor this code I'd probably make it explicit so that the 'else' clause was always some generic unknown operation response. This way if the batch api adds more operation types in the future, it would be clear that it wasn't supported in gitea (yet) instead of being assumed to be the 'download' operation.
... I did not do (2) while I was here because I just wanted to get this parallelized with minimal refactoring first.